extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessor.java

```diff
@@ -146,18 +147,30 @@ protected SegmentReference mapSegment(final Segment segment)
 
   private boolean initializeSegmentMapFn(final IntSet readableInputs)
   {
+    final AtomicLong cpuAccumulator = new AtomicLong();
     if (segmentMapFn != null) {
       return true;
     } else if (broadcastJoinHelper == null) {
       segmentMapFn = Function.identity();
       return true;
     } else {
       final boolean retVal = broadcastJoinHelper.buildBroadcastTablesIncrementally(readableInputs);
-
+      DataSource inlineChannelDataSource = broadcastJoinHelper.inlineChannelData(query.getDataSource());
       if (retVal) {
-        segmentMapFn = broadcastJoinHelper.makeSegmentMapFn(query);
+        if (inlineChannelDataSource instanceof InputNumberDataSource) {
+          InputNumberDataSource inputNumberDataSource = (InputNumberDataSource) query.getDataSource();
+          // The InputNumberDataSource requires a BroadcastJoinHelper to be able to create its
+          // segment map function.  It would be a lot better if the InputNumberDataSource actually
+          // had a way to get that injected into it on its own, but the relationship between these objects
+          // was figured out during a refactor and using a setter here seemed like the least-bad way to
+          // make progress on the refactor without breaking functionality.  Hopefully, some future
+          // developer will move this away from a setter.
+          inputNumberDataSource.setBroadcastJoinHelper(broadcastJoinHelper);
+          segmentMapFn = inputNumberDataSource.createSegmentMapFunction(query, cpuAccumulator);
+        } else {
+          segmentMapFn = inlineChannelDataSource.createSegmentMapFunction(query, cpuAccumulator);
```

Comment:
@abhishekagarwal87 I have updated this part. Should address your earlier concern

Reply:
Did we find a test that was broken with the previous code?  If not, then please revert back to the code that @abhishekagarwal87 initially commented on and keep that code active.  That code is not broken until we can actually reproduce the breakage.

Reply:
This example passes with the new code changes. The problem discovered was :
1. There was a lack of a proper test case for this to be discovered. I am adding the test suggested here 
2. In the previous case, the identity function in the last else part was incorrect as it would not create the correct segment function for this example. This has been now updated by calling `createSegmentMapFunction` on the correct data source.
3. All unit tests for MSQ, CalciteJoins now complete successfully